src/zap.c

```diff
@@ -5241,174 +5326,60 @@ int osym, dmgtyp;
-                destroy_one_item(obj, osym, dmgtyp);
-                break;
-            }
+    pline("elig stacks = %d dmg = %d, limit = %d", elig_stacks, dmg_in, limit);
+    if (elig_stacks > limit) {
+        elig_stacks = limit; /* so we can loop up to elig_stacks */
     }
-    return;
-}
-
-int
-destroy_mitem(mtmp, osym, dmgtyp)
-struct monst *mtmp;
-int osym, dmgtyp;
-{
-    struct obj *obj;
-    int skip, tmp = 0;
-    long i, cnt, quan;
-    int dindx;
-    boolean vis;
-
-    if (mtmp == &g.youmonst) { /* this simplifies artifact_hit() */
-        destroy_item(osym, dmgtyp);
-        return 0; /* arbitrary; value doesn't matter to artifact_hit() */
-    }
-
-    vis = canseemon(mtmp);
-
-    /* see destroy_item(); object destruction could disrupt inventory list */
-    bypass_objlist(mtmp->minvent, FALSE); /* clear bypass bit for minvent */
-
-    while ((obj = nxt_unbypassed_obj(mtmp->minvent)) != 0) {
-        if (obj->oclass != osym)
-            continue; /* test only objs of type osym */
-        skip = 0;
-        quan = 0L;
-        dindx = 0;
-
-        switch (dmgtyp) {
-        case AD_COLD:
-            if (osym == POTION_CLASS && obj->otyp != POT_OIL) {
-                quan = obj->quan;
-                dindx = 0;
-                tmp++;
-            } else
-                skip++;
-            break;
-        case AD_FIRE:
-            if (obj->otyp == SCR_FIRE || obj->otyp == SPE_FIREBALL)
-                skip++;
-            if (obj->otyp == SPE_BOOK_OF_THE_DEAD) {
-                skip++;
-                if (vis)
-                    pline("%s glows a strange %s, but remains intact.",
-                          The(distant_name(obj, xname)), hcolor("dark red"));
-            }
-            quan = obj->quan;
-            switch (osym) {
-            case POTION_CLASS:
-                dindx = (obj->otyp != POT_OIL) ? 1 : 2;
-                tmp++;
-                break;
-            case SCROLL_CLASS:
-                dindx = 3;
-                tmp++;
-                break;
-            case SPBOOK_CLASS:
-                dindx = 4;
-                tmp++;
-                break;
-            case FOOD_CLASS:
-                if (obj->otyp == GLOB_OF_GREEN_SLIME) {
-                    dindx = 1; /* boil and explode */
-                    tmp += (obj->owt + 19) / 20;
-                } else {
-                    skip++;
+    for (defer = 0; defer <= 1; ++defer) {
+        /* if we saved some items for later (most likely just a worn ring
+           of levitation) and they're still in inventory, handle them on the
+           second iteration of the loop */
+        struct obj *nobj;
+        for (obj = *objchn; obj; obj = nobj) {
+            nobj = obj->nobj;
+            for (i = 0; i < elig_stacks; ++i) {
+                if (obj->o_id == items_to_destroy[i].oid
+                    && (items_to_destroy[i].deferred == (defer == 1))) {
+                    dmg_out += maybe_destroy_item(mon, obj, dmgtyp);
```

Comment:
I guess a `break` could be added here so that once the object is found in `items_to_destroy` the remaining indices aren't looped through and checked needlessly (an item can't end up on the list twice, can it?), but maybe not all that important or useful with only 20 items max.